semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala

```diff
@@ -1,86 +1,64 @@
-package scala.meta.internal.metacp
+package scala.meta.internal.javacp
 
 import java.nio.file.Files
 import java.nio.file.Path
-
 import scala.collection.JavaConverters._
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 import scala.collection.mutable.ListBuffer
-import scala.meta.internal.metacp.asm.JavaTypeSignature._
-import scala.meta.internal.metacp.asm._
+import scala.meta.internal.javacp.asm._
+import scala.meta.internal.metacp._
 import scala.meta.internal.semanticdb3.SymbolInformation.{Kind => k}
 import scala.meta.internal.{semanticdb3 => s}
-import scala.tools.asm.ClassReader
 import scala.tools.asm.tree.ClassNode
 import scala.tools.asm.tree.FieldNode
 import scala.tools.asm.tree.InnerClassNode
 import scala.tools.asm.tree.MethodNode
 import scala.tools.asm.{Opcodes => o}
 
 object Javacp {
-
-  def sinfos(
-      root: Path,
-      file: Path,
-      isVisited: mutable.Set[Path]
-  ): Seq[s.SymbolInformation] = {
-    sinfosFromOuterClass(root, file, 0, Scope.empty, isVisited)
-  }
-
-  def sinfosFromOuterClass(
-      root: Path,
-      file: Path,
-      outerClassAccess: Int,
-      scope: Scope,
-      isVisited: mutable.Set[Path]): Seq[s.SymbolInformation] = {
-    if (isVisited(file)) Nil
-    else {
-      isVisited += file
-      val bytes = Files.readAllBytes(file)
-      val node = parseClassNode(bytes)
-      if (isAnonymousClass(node)) {
-        // Skip anonymous classes like we do for Scala symbols.
-        Nil
-      } else {
-        sinfosClassNode(node, outerClassAccess, scope, root, isVisited)
-      }
+  def parse(classfile: ToplevelClassfile): Option[ToplevelInfos] = {
+    sinfos(classfile, classfile.node, 0, Scope.empty) match {
+      case others :+ toplevel =>
+        Some(ToplevelInfos(classfile, List(toplevel), others.toList))
+      case _ =>
+        None
     }
   }
 
-  def sinfosClassNode(
+  private def sinfos(
+      toplevel: ToplevelClassfile,
       node: ClassNode,
-      outerClassAccess: Int,
-      scope: Scope,
-      root: Path,
-      isVisited: mutable.Set[Path]): Seq[s.SymbolInformation] = {
+      access: Int,
```

Comment:
I intentionally renamed this from access to outerClassAccess in hope to make it clearer. No objections from me if you prefer it this way.

Reply:
I think that `outerClassAccess` is somewhat misleading. It's not `access` of an outer class, but `access` of an `InnerClass` entry in outer class. I don't have better ideas though, so let's use `access` for the time being if you don't mind.

Reply:
I agree.